@Zheng-Bicheng 那我目前这边测例无法通过就不用改了吗，请麻烦看一下还有哪里需要修改。

目前单测报的是版本不支持的问题：

```
[Paddle2ONNX] Start to parse PaddlePaddle model...
[Paddle2ONNX] Model file path: deform_conv2d/cliped_model.pdmodel
[Paddle2ONNX] Parameters file path: 
[Paddle2ONNX] Start to parsing Paddle model...
[ERROR] Paddle2ONNX now only support opset version in range of [7, 18].


--------------------------------------
C++ Traceback (most recent call last):
--------------------------------------
0   paddle2onnx::Export(char const*, char const*, char**, int*, int, bool, bool, bool, bool, bool, paddle2onnx::CustomOp*, int, char const*, char**, int*, char const*, bool*, bool, char**, int)

----------------------
Error Message Summary:
----------------------
FatalError: `Process abort signal` is detected by the operating system.
  [TimeInfo: *** Aborted at 1716473326 (unix time) try "date -d @1716473326" if you are using GNU date ***]
  [SignalInfo: *** SIGABRT (@0x7b3da) received by PID 504794 (TID 0x7f2e08d8e4c0) from PID 504794 ***]

[1]    504794 abort (core dumped)  python tests/test_deform_conv2d.py
```

--- a/paddle2onnx/mapper/nn/deform_conv2d.cc
+++ b/paddle2onnx/mapper/nn/deform_conv2d.cc
@@ -1,4 +1,4 @@
-// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
+// Copyright (c) 2024 PaddlePaddle Authors. All Rights Reserved.
 //
 // Licensed under the Apache License, Version 2.0 (the "License");
 // you may not use this file except in compliance with the License.

--- a/paddle2onnx/mapper/nn/deform_conv2d.cc
+++ b/paddle2onnx/mapper/nn/deform_conv2d.cc
@@ -23,21 +23,16 @@ namespace paddle2onnx
 
   int32_t DeformConv2dMapper::GetMinOpset(bool verbose)
   {
-    return 7;
+    return 19;
   }
 
-  void DeformConv2dMapper::Opset7()
+  void DeformConv2dMapper::Opset19()
   {
     auto kernel_info = GetInput("Filter");
     auto input_info = GetInput("Input");
     auto offset_info = GetInput("Offset");
     auto mask_info = GetInput("Mask");
     auto output_info = GetOutput("Output");
-
-    // TODO
-    // 1. 这个变量名称映射paddle的是哪个呢？看上去是ops.yaml和op_compat.yaml吗？
-    // 2. c++ api上没有bias，这个是def deformable_conv调用C_ops后再加上的，这里怎么处理呢？optional的输入也要全写入吗
-    // 3. 这个opsetx怎么具体对应起来呢？跟onnx似乎对不上
     auto node = helper_->MakeNode(
         "DeformConv", {input_info[0].name, kernel_info[0].name, offset_info[0].name}, {output_info[0].name});
 

--- a/paddle2onnx/mapper/nn/deform_conv2d.cc
+++ b/paddle2onnx/mapper/nn/deform_conv2d.cc
@@ -1,0 +1,53 @@
+// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle2onnx/mapper/nn/deform_conv2d.h"
+
+#include <string>
+#include <vector>
+
+namespace paddle2onnx
+{
+  REGISTER_MAPPER(deformable_conv, DeformConv2dMapper)
+
+  int32_t DeformConv2dMapper::GetMinOpset(bool verbose)
+  {
+    return 7;
+  }
+
+  void DeformConv2dMapper::Opset7()
+  {
+    auto kernel_info = GetInput("Filter");
+    auto input_info = GetInput("Input");
+    auto offset_info = GetInput("Offset");
+    auto mask_info = GetInput("Mask");
+    auto output_info = GetOutput("Output");
+
+    // TODO
+    // 1. 这个变量名称映射paddle的是哪个呢？看上去是ops.yaml和op_compat.yaml吗？
+    // 2. c++ api上没有bias，这个是def deformable_conv调用C_ops后再加上的，这里怎么处理呢？optional的输入也要全写入吗
+    // 3. 这个opsetx怎么具体对应起来呢？跟onnx似乎对不上
+    auto node = helper_->MakeNode(
+        "DeformConv", {input_info[0].name, kernel_info[0].name, offset_info[0].name}, {output_info[0].name});
+
+    AddAttribute(node, "dilations", dilations_);
+    AddAttribute(node, "group", groups_);
+    std::vector<int64_t> kernel_shape = {kernel_info[0].shape[2],
+                                         kernel_info[0].shape[3]};
+    AddAttribute(node, "kernel_shape", kernel_shape);
+    AddAttribute(node, "pads", paddings_);
+    AddAttribute(node, "strides", strides_);
+  }
+
+} // namespace paddle2onnx

--- a/paddle2onnx/mapper/nn/deform_conv2d.h
+++ b/paddle2onnx/mapper/nn/deform_conv2d.h
@@ -18,31 +18,34 @@
 
 #include "paddle2onnx/mapper/mapper.h"
 
-namespace paddle2onnx {
+namespace paddle2onnx
+{
 
-class DeformConv2dMapper : public Mapper {
- public:
-  DeformConv2dMapper(const PaddleParser& p, OnnxHelper* helper, int64_t block_id,
-               int64_t op_id)
-      : Mapper(p, helper, block_id, op_id) {
-    GetAttr("strides", &strides_);
-    GetAttr("paddings", &paddings_);
-    GetAttr("dilations", &dilations_);
-    GetAttr("deformable_groups", &deformable_groups_);
-    GetAttr("groups", &groups_);
-    GetAttr("im2col_step", &im2col_step_);
-  }
+  class DeformConv2dMapper : public Mapper
+  {
+  public:
+    DeformConv2dMapper(const PaddleParser &p, OnnxHelper *helper, int64_t block_id,
+                       int64_t op_id)
+        : Mapper(p, helper, block_id, op_id)
+    {
+      GetAttr("strides", &strides_);
+      GetAttr("paddings", &paddings_);
+      GetAttr("dilations", &dilations_);
+      GetAttr("deformable_groups", &deformable_groups_);
+      GetAttr("groups", &groups_);
+      GetAttr("im2col_step", &im2col_step_);
+    }
 
-  int32_t GetMinOpset(bool verbose = false);
-  void Opset19();
+    int32_t GetMinOpset(bool verbose = false);
+    void Opset19();
 
- private:
-  std::vector<int64_t> strides_;
-  std::vector<int64_t> paddings_;
-  std::vector<int64_t> dilations_;
-  int64_t deformable_groups_;
-  int64_t groups_;
-  int64_t im2col_step_;
-};
+  private:
+    std::vector<int64_t> strides_;
+    std::vector<int64_t> paddings_;
+    std::vector<int64_t> dilations_;
+    int64_t deformable_groups_;
+    int64_t groups_;
+    int64_t im2col_step_;
+  };
 
-}  // namespace paddle2onnx
+} // namespace paddle2onnx

--- a/paddle2onnx/mapper/nn/deform_conv2d.h
+++ b/paddle2onnx/mapper/nn/deform_conv2d.h
@@ -34,7 +34,7 @@ class DeformConv2dMapper : public Mapper {
   }
 
   int32_t GetMinOpset(bool verbose = false);
-  void Opset7();
+  void Opset19();
 
  private:
   std::vector<int64_t> strides_;

--- a/paddle2onnx/mapper/nn/deform_conv2d.h
+++ b/paddle2onnx/mapper/nn/deform_conv2d.h
@@ -1,0 +1,48 @@
+// Copyright (c) 2024 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#include <string>
+#include <vector>
+
+#include "paddle2onnx/mapper/mapper.h"
+
+namespace paddle2onnx {
+
+class DeformConv2dMapper : public Mapper {
+ public:
+  DeformConv2dMapper(const PaddleParser& p, OnnxHelper* helper, int64_t block_id,
+               int64_t op_id)
+      : Mapper(p, helper, block_id, op_id) {
+    GetAttr("strides", &strides_);
+    GetAttr("paddings", &paddings_);
+    GetAttr("dilations", &dilations_);
+    GetAttr("deformable_groups", &deformable_groups_);
+    GetAttr("groups", &groups_);
+    GetAttr("im2col_step", &im2col_step_);
+  }
+
+  int32_t GetMinOpset(bool verbose = false);
+  void Opset7();
+
+ private:
+  std::vector<int64_t> strides_;
+  std::vector<int64_t> paddings_;
+  std::vector<int64_t> dilations_;
+  int64_t deformable_groups_;
+  int64_t groups_;
+  int64_t im2col_step_;
+};
+
+}  // namespace paddle2onnx

--- a/tests/test_deform_conv2d.py
+++ b/tests/test_deform_conv2d.py
@@ -21,12 +21,13 @@ class Net(paddle.nn.Layer):
 def test_deform_conv2d():
     """
     api: paddle.vision.ops.deform_conv2d
-    op version: 7
+    op version: 19
     """
+
     op = Net()
     op.eval()
     # net, name, ver_list, delta=1e-6, rtol=1e-5
-    obj = APIOnnx(op, "deform_conv2d", [7])
+    obj = APIOnnx(op, "deform_conv2d", [19])
     kh, kw = 3, 3
     input = paddle.rand((8, 1, 28, 28))
     offset = paddle.rand((8, 2 * kh * kw, 26, 26))
@@ -35,36 +36,6 @@ def test_deform_conv2d():
     obj.set_input_data("input_data", input, offset, weight, mask)
     obj.run()
 
-
-"""
-测例报错
-/root/miniconda3/envs/paddle_onnx/lib/python3.9/site-packages/paddle/static/io.py:610: UserWarning: no variable in your model, please ensure there are any variables in your model to save
-  warnings.warn(
-[Paddle2ONNX] Start to parse PaddlePaddle model...
-[Paddle2ONNX] Model file path: deform_conv2d/cliped_model.pdmodel
-[Paddle2ONNX] Parameters file path: 
-[Paddle2ONNX] Start to parsing Paddle model...
-[Paddle2ONNX] Use opset_version = 7 for ONNX export.
-[Paddle2ONNX] The exported ONNX model is invalid.
-[Paddle2ONNX] Model checker error log: No Op registered for DeformConv with domain_version of 7
-
-==> Context: Bad node spec for node. Name: p2o.DeformConv.0 OpType: DeformConv
-[Paddle2ONNX] PaddlePaddle model is exported as ONNX format now.
-Traceback (most recent call last):
-  File "/wuzp/Paddle2ONNX/tests/test_deform_conv2d.py", line 40, in <module>
-    test_deform_conv2d()
-  File "/wuzp/Paddle2ONNX/tests/test_deform_conv2d.py", line 36, in test_deform_conv2d
-    obj.run()
-  File "/wuzp/Paddle2ONNX/tests/onnxbase.py", line 420, in run
-    res_fict[str(v)] = self._mk_onnx_res(ver=v)
-  File "/wuzp/Paddle2ONNX/tests/onnxbase.py", line 300, in _mk_onnx_res
-    sess = InferenceSession(
-  File "/root/miniconda3/envs/paddle_onnx/lib/python3.9/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 419, in __init__
-    self._create_inference_session(providers, provider_options, disabled_optimizers)
-  File "/root/miniconda3/envs/paddle_onnx/lib/python3.9/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 472, in _create_inference_session
-    sess = C.InferenceSession(session_options, self._model_path, True, self._read_config_from_model)
-onnxruntime.capi.onnxruntime_pybind11_state.InvalidGraph: [ONNXRuntimeError] : 10 : INVALID_GRAPH : Load model from /wuzp/Paddle2ONNX/deform_conv2d/deform_conv2d_7.onnx failed:This is an invalid model. In Node, ("p2o.DeformConv.0", DeformConv, "", -1) : ("0": tensor(float),"2": tensor(float),"1": tensor(float),) -> ("save_infer_model/scale_0.tmp_0": tensor(float),) , Error No Op registered for DeformConv with domain_version of 7
-"""
 
 if __name__ == "__main__":
     test_deform_conv2d()

--- a/tests/test_deform_conv2d.py
+++ b/tests/test_deform_conv2d.py
@@ -1,0 +1,70 @@
+import paddle
+from onnxbase import APIOnnx, randtool
+
+
+class Net(paddle.nn.Layer):
+    """
+    simple Net
+    """
+
+    def __init__(self):
+        super(Net, self).__init__()
+
+    def forward(self, inputs, offset, weight, mask):
+        """
+        forward
+        """
+        x = paddle.vision.ops.deform_conv2d(inputs, offset, weight, mask=mask)
+        return x
+
+
+def test_deform_conv2d():
+    """
+    api: paddle.vision.ops.deform_conv2d
+    op version: 7
+    """
+    op = Net()
+    op.eval()
+    # net, name, ver_list, delta=1e-6, rtol=1e-5
+    obj = APIOnnx(op, "deform_conv2d", [7])
+    kh, kw = 3, 3
+    input = paddle.rand((8, 1, 28, 28))
+    offset = paddle.rand((8, 2 * kh * kw, 26, 26))
+    mask = paddle.rand((8, kh * kw, 26, 26))
+    weight = paddle.rand((16, 1, kh, kw))
+    obj.set_input_data("input_data", input, offset, weight, mask)
+    obj.run()
+
+
+"""
+测例报错
+/root/miniconda3/envs/paddle_onnx/lib/python3.9/site-packages/paddle/static/io.py:610: UserWarning: no variable in your model, please ensure there are any variables in your model to save
+  warnings.warn(
+[Paddle2ONNX] Start to parse PaddlePaddle model...
+[Paddle2ONNX] Model file path: deform_conv2d/cliped_model.pdmodel
+[Paddle2ONNX] Parameters file path: 
+[Paddle2ONNX] Start to parsing Paddle model...
+[Paddle2ONNX] Use opset_version = 7 for ONNX export.
+[Paddle2ONNX] The exported ONNX model is invalid.
+[Paddle2ONNX] Model checker error log: No Op registered for DeformConv with domain_version of 7
+
+==> Context: Bad node spec for node. Name: p2o.DeformConv.0 OpType: DeformConv
+[Paddle2ONNX] PaddlePaddle model is exported as ONNX format now.
+Traceback (most recent call last):
+  File "/wuzp/Paddle2ONNX/tests/test_deform_conv2d.py", line 40, in <module>
+    test_deform_conv2d()
+  File "/wuzp/Paddle2ONNX/tests/test_deform_conv2d.py", line 36, in test_deform_conv2d
+    obj.run()
+  File "/wuzp/Paddle2ONNX/tests/onnxbase.py", line 420, in run
+    res_fict[str(v)] = self._mk_onnx_res(ver=v)
+  File "/wuzp/Paddle2ONNX/tests/onnxbase.py", line 300, in _mk_onnx_res
+    sess = InferenceSession(
+  File "/root/miniconda3/envs/paddle_onnx/lib/python3.9/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 419, in __init__
+    self._create_inference_session(providers, provider_options, disabled_optimizers)
+  File "/root/miniconda3/envs/paddle_onnx/lib/python3.9/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 472, in _create_inference_session
+    sess = C.InferenceSession(session_options, self._model_path, True, self._read_config_from_model)
+onnxruntime.capi.onnxruntime_pybind11_state.InvalidGraph: [ONNXRuntimeError] : 10 : INVALID_GRAPH : Load model from /wuzp/Paddle2ONNX/deform_conv2d/deform_conv2d_7.onnx failed:This is an invalid model. In Node, ("p2o.DeformConv.0", DeformConv, "", -1) : ("0": tensor(float),"2": tensor(float),"1": tensor(float),) -> ("save_infer_model/scale_0.tmp_0": tensor(float),) , Error No Op registered for DeformConv with domain_version of 7
+"""
+
+if __name__ == "__main__":
+    test_deform_conv2d()

